flocker/node/agents/ebs.py

```diff
@@ -390,122 +410,221 @@ def ec2_client(region, zone, access_key_id, secret_access_key):
     :return: An ``_EC2`` giving information about EC2 client connection
         and EC2 instance zone.
     """
-    # Set 2 retry knobs in Boto to BOTO_NUM_RETRIES:
-    # 1. ``num_retries``:
-    # Request automatic exponential backoff and retry
-    # attempts by Boto if an EC2 API call fails with
-    # ``RequestLimitExceeded`` due to system load.
-    # 2. ``metadata_service_num_attempts``:
-    # Request for retry attempts by Boto to
-    # retrieve data from Metadata Service used to retrieve
-    # credentials for IAM roles on EC2 instances.
-    if not config.has_section('Boto'):
-        config.add_section('Boto')
-    config.set('Boto', 'num_retries', BOTO_NUM_RETRIES)
-    config.set('Boto', 'metadata_service_num_attempts', BOTO_NUM_RETRIES)
-
-    # Get Boto EC2 connection with ``EC2ResponseError`` logged by Eliot.
-    connection = ec2.connect_to_region(region,
-                                       aws_access_key_id=access_key_id,
-                                       aws_secret_access_key=secret_access_key)
-    return _EC2(zone=zone,
-                connection=_LoggedBotoConnection(connection=connection))
-
-
-def _boto_logged_method(method_name, original_name):
+    connection = boto3.session.Session(
+        aws_access_key_id=access_key_id,
+        aws_secret_access_key=secret_access_key
+    )
+    connection._session.set_config_variable(
+        'metadata_service_num_attempts', BOTO_NUM_RETRIES)
+    ec2_resource = connection.resource("ec2", region_name=region)
+    return EC2Client(zone=zone, connection=ec2_resource)
+
+
+def boto3_log(method):
     """
-    Run a boto.ec2.connection.EC2Connection method and
+    Decorator to run a boto3.ec2.ServiceResource method and
     log additional information about any exceptions that are raised.
 
-    :param str method_name: The name of the method of the wrapped object to
-        call.
-    :param str original_name: The name of the attribute of self where the
-        wrapped object can be found.
+    :param func method_name: The method to call.
 
-    :return: A function which will call the method of the wrapped object and do
+    :return: A function which will call the method and do
         the extra exception logging.
     """
-    def _run_with_logging(self, *args, **kwargs):
+    def _run_with_logging(*args, **kwargs):
         """
-        Run given boto.ec2.connection.EC2Connection method with exception
-        logging for ``EC2ResponseError``.
+        Run given boto3.ec2.ServiceResource method with exception
+        logging for ``ClientError``.
         """
-        original = getattr(self, original_name)
-        method = getattr(original, method_name)
+        def log_boto3_error(e):
```

Comment:
When an exception is raised inside an Eliot action context and not caught the exception is automatically logged. And in latest version of Eliot, as used by Flocker, you can register additional field extraction for specific exceptions.  So:

1. On module level register an exception extractor that extracts those fields from `ClientError` (see http://eliot.readthedocs.org/en/0.10.1/generating/errors.html#custom-exception-logging for details).
2. You can then get rid of `BOTO_EC2RESPONSE_ERROR` and the need to catch and reraise `ClientError`.